From eac0ed54414904107d0b9d723ef5c2a8d8d3b655 Mon Sep 17 00:00:00 2001
From: Jean-Michel Bachot <jean-michel.bachot@intel.com>
Date: Tue, 16 Jul 2013 17:36:08 +0200
Subject: [PORT FROM MAIN] wifi/P2P: ensure virtual group interface removal from the driver upon P2P deinitialize

BZ: 99166 111626

While Deinitialize per-interface P2P data, remove group interface
from the driver if this wpa_s instance is on top of a P2P group interface.
This ensures p2p group virtual interface removal in driver upon wpa_supplicant
closure.

Change-Id: I2e53428f5eb1268ae417ddf8bf87213884fc3790
Orig-Change-Id: I8e42c7356282610e0c75aba81d7055b73ff05bb4
Category: device enablement
Domain: CWS.WIFI-Common
Origin: upstream
Upstream-Candidate: yes
Signed-off-by: Jean-Michel Bachot <jean-michel.bachot@intel.com>
---
 wpa_supplicant/p2p_supplicant.c |   17 ++++++++++++++++-
 1 files changed, 16 insertions(+), 1 deletions(-)

diff --git a/wpa_supplicant/p2p_supplicant.c b/wpa_supplicant/p2p_supplicant.c
index 972d61e..b214488 100644
--- a/wpa_supplicant/p2p_supplicant.c
+++ b/wpa_supplicant/p2p_supplicant.c
@@ -3433,8 +3433,23 @@ void wpas_p2p_deinit(struct wpa_supplicant *wpa_s)
 	eloop_cancel_timeout(wpas_p2p_group_idle_timeout, wpa_s, NULL);
 	wpas_p2p_remove_pending_group_interface(wpa_s);
 
-	/* TODO: remove group interface from the driver if this wpa_s instance
+	/* remove group interface from the driver if this wpa_s instance
 	 * is on top of a P2P group interface */
+	if (wpa_s->p2p_group_interface != NOT_P2P_GROUP_INTERFACE) {
+		struct wpa_global *global;
+		char *ifname;
+		enum wpa_driver_if_type type;
+		wpa_printf(MSG_DEBUG, "P2P: Remove group interface %s",
+			wpa_s->ifname);
+		global = wpa_s->global;
+		ifname = os_strdup(wpa_s->ifname);
+		type = wpas_p2p_if_type(wpa_s->p2p_group_interface);
+		wpa_supplicant_remove_iface(wpa_s->global, wpa_s, 0);
+		wpa_s = global->ifaces;
+		if (wpa_s && ifname)
+			wpa_drv_if_remove(wpa_s, type, ifname);
+		os_free(ifname);
+	}
 }
 
 
-- 
1.7.4.1

